File: src/main/java/com/verlumen/tradestream/backtesting/BacktestRunnerImpl.java
package com.verlumen.tradestream.backtesting;

import static com.google.common.base.Preconditions.checkArgument;

import com.google.common.collect.ImmutableList;
import com.google.inject.Inject;
import com.google.protobuf.InvalidProtocolBufferException;
import com.verlumen.tradestream.strategies.StrategyManager;
import com.verlumen.tradestream.ta4j.BarSeriesFactory;
import org.ta4j.core.AnalysisCriterion;
import org.ta4j.core.BarSeries;
import org.ta4j.core.BaseTradingRecord;
import org.ta4j.core.Strategy;
import org.ta4j.core.TradingRecord;
import org.ta4j.core.criteria.pnl.ReturnCriterion;
import org.ta4j.core.criteria.pnl.ProfitLossCriterion;
import org.ta4j.core.criteria.pnl.ProfitLossRatioCriterion;
import org.ta4j.core.num.Num;
import java.util.ArrayList;
import java.util.List;

/**
 * Implementation of BacktestRunner that evaluates trading strategies using Ta4j.
 */
final class BacktestRunnerImpl implements BacktestRunner {
    private final BarSeriesFactory barSeriesFactory;
    private final StrategyManager strategyManager;

    @Inject
    BacktestRunnerImpl(
        BarSeriesFactory barSeriesFactory,
        StrategyManager strategyManager) {
        this.barSeriesFactory = barSeriesFactory;
        this.strategyManager = strategyManager;
    }

    @Override
    public BacktestResult runBacktest(BacktestRequest request) throws InvalidProtocolBufferException {
        checkArgument(request.getCandlesList().size() > 0, "Bar series cannot be empty");

        BarSeries series = barSeriesFactory.createBarSeries(
            ImmutableList.copyOf(request.getCandlesList())
        );
        Strategy strategy = strategyManager.createStrategy(
            series,  request.getStrategy().getType(),request.getStrategy().getParameters()
        );
                
        // Run the strategy
        TradingRecord tradingRecord = runStrategy(series, strategy);

        // Calculate basic metrics
        double cumulativeReturn = calculateMetric(series, tradingRecord, 
            new ProfitLossCriterion());
        
        double profitFactor = calculateMetric(series, tradingRecord,
            new ProfitLossRatioCriterion());
        
        double annualizedReturn = calculateAnnualizedReturn(series, tradingRecord);
        
        // Calculate risk metrics
        double volatility = calculateVolatility(series);
        double maxDrawdown = calculateMaxDrawdown(series);
        
        // Trade statistics
        int numberOfTrades = tradingRecord.getPositions().size();
        double winRate = calculateWinRate(series, tradingRecord);
        double averageTradeDuration = calculateAverageTradeDuration(tradingRecord);

        // Risk-adjusted returns
        double sharpeRatio = calculateSharpeRatio(cumulativeReturn, volatility);
        double sortinoRatio = calculateSortinoRatio(series, tradingRecord);

        // Alpha/Beta (simplified calculation)
        double alpha = 0.0; // TODO: Implement when benchmark data is available
        double beta = 1.0;  // TODO: Implement when benchmark data is available

        double score = calculateScore(sharpeRatio, maxDrawdown, winRate, annualizedReturn, profitFactor);

        return BacktestResult.newBuilder()
            .setCumulativeReturn(cumulativeReturn)
            .setAnnualizedReturn(annualizedReturn)
            .setSharpeRatio(sharpeRatio)
            .setSortinoRatio(sortinoRatio)
            .setMaxDrawdown(maxDrawdown)
            .setVolatility(volatility)
            .setWinRate(winRate)
            .setProfitFactor(profitFactor)
            .setNumberOfTrades(numberOfTrades)
            .setAverageTradeDuration(averageTradeDuration)
            .setAlpha(alpha)
            .setBeta(beta)
            .setStrategyScore(score)
            .build();
    }

    private TradingRecord runStrategy(BarSeries series, Strategy strategy) {
        TradingRecord tradingRecord = new BaseTradingRecord();
        
        // Skip unstable period at the start
        for (int i = strategy.getUnstableBars(); i < series.getBarCount(); i++) {
            // Check if we should enter long position
            if (strategy.shouldEnter(i)) {
                // Enter with a position size of 1 unit
                tradingRecord.enter(i, series.getBar(i).getClosePrice(), series.numOf(1));
            }
            // Check if we should exit an open position
            else if (strategy.shouldExit(i) && tradingRecord.getCurrentPosition().isOpened()) {
                tradingRecord.exit(i, series.getBar(i).getClosePrice(), series.numOf(1));
            }
        }
        
        return tradingRecord;
    }

    private double calculateMetric(BarSeries series, TradingRecord record,
            AnalysisCriterion criterion) {
        return criterion.calculate(series, record).doubleValue();
    }

    private double calculateVolatility(BarSeries series) {
        if (series.getBarCount() < 2) {
            return 0.0;
        }

        List<Double> returns = new ArrayList<>();
        for (int i = 1; i < series.getBarCount(); i++) {
            Num previousClose = series.getBar(i - 1).getClosePrice();
            Num currentClose = series.getBar(i).getClosePrice();
            double dailyReturn = currentClose.minus(previousClose)
                .dividedBy(previousClose)
                .doubleValue();
            returns.add(dailyReturn);
        }
        
        // Calculate standard deviation
        double mean = returns.stream()
            .mapToDouble(Double::doubleValue)
            .average()
            .orElse(0.0);
            
        double variance = returns.stream()
            .mapToDouble(r -> Math.pow(r - mean, 2))
            .average()
            .orElse(0.0);
            
        return Math.sqrt(variance);
    }

    private double calculateMaxDrawdown(BarSeries series) {
        if (series.getBarCount() == 0) {
            return 0.0;
        }

        double maxDrawdown = 0.0;
        double peak = series.getBar(0).getClosePrice().doubleValue();
        
        for (int i = 1; i < series.getBarCount(); i++) {
            double price = series.getBar(i).getClosePrice().doubleValue();
            
            if (price > peak) {
                peak = price;
            }
            
            double drawdown = (peak - price) / peak;
            maxDrawdown = Math.max(maxDrawdown, drawdown);
        }
        
        return maxDrawdown;
    }

    private double calculateWinRate(BarSeries series, TradingRecord record) {
        if (record.getPositions().isEmpty()) {
            return 0.0;
        }

        long winningTrades = record.getPositions().stream()
            .filter(position -> position.isClosed() && 
                position.getProfit().isPositive())
            .count();
        
        return (double) winningTrades / record.getPositions().size();
    }

    private double calculateAverageTradeDuration(TradingRecord record) {
        if (record.getPositions().isEmpty()) {
            return 0.0;
        }

        double totalDuration = record.getPositions().stream()
            .filter(position -> position.isClosed())
            .mapToInt(position -> position.getExit().getIndex() - position.getEntry().getIndex())
            .sum();
        
        long closedPositions = record.getPositions().stream()
            .filter(position -> position.isClosed())
            .count();
        
        return closedPositions > 0 ? totalDuration / closedPositions : 0.0;
    }

    private double calculateAnnualizedReturn(BarSeries series, TradingRecord record) {
        double totalReturn = calculateMetric(series, record, new ReturnCriterion());
        int barsPerYear = 252 * 1440; // Assuming 1-minute bars and 252 trading days
        double years = (double) series.getBarCount() / barsPerYear;
        
        // Use compound annual growth rate formula
        return Math.pow(1 + totalReturn, 1 / years) - 1;
    }

    private double calculateSharpeRatio(double returns, double volatility) {
        double riskFreeRate = 0.02; // Assume 2% risk-free rate
        return volatility == 0 ? 0 : (returns - riskFreeRate) / volatility;
    }

    private double calculateSortinoRatio(BarSeries series, TradingRecord record) {
        // Simplified Sortino calculation using only negative returns
        List<Double> negativeReturns = new ArrayList<>();
        
        for (int i = 1; i < series.getBarCount(); i++) {
            double previousPrice = series.getBar(i - 1).getClosePrice().doubleValue();
            double currentPrice = series.getBar(i).getClosePrice().doubleValue();
            double return_ = (currentPrice - previousPrice) / previousPrice;
            
            if (return_ < 0) {
                negativeReturns.add(return_);
            }
        }
        
        if (negativeReturns.isEmpty()) {
            return 0.0;
        }
        
        // Calculate downside deviation
        double meanNegativeReturn = negativeReturns.stream()
            .mapToDouble(Double::doubleValue)
            .average()
            .orElse(0.0);
            
        double downsideDeviation = Math.sqrt(
            negativeReturns.stream()
                .mapToDouble(r -> Math.pow(r - meanNegativeReturn, 2))
                .average()
                .orElse(0.0)
        );
        
        // Calculate Sortino ratio using total return
        double totalReturn = calculateMetric(series, record, new ReturnCriterion());
        double riskFreeRate = 0.02; // Assume 2% risk-free rate
        
        return downsideDeviation == 0 ? 0 : (totalReturn - riskFreeRate) / downsideDeviation;
    }

    private double calculateScore(double sharpeRatio, double maxDrawdown, double winRate, double annualizedReturn, double profitFactor) {
        return 
            0.25 * normalize(sharpeRatio) +          // Risk-adjusted returns
            0.20 * (1 - maxDrawdown) +              // Capital preservation
            0.20 * winRate +                        // Consistency
            0.20 * normalize(annualizedReturn) +    // Absolute returns
            0.15 * normalize(profitFactor);         // Profit efficiency
    }

    private double normalize(double value) {
        // Simple min-max normalization with reasonable bounds
        double min = -1.0;
        double max = 2.0;
        return Math.max(0, Math.min(1, (value - min) / (max - min)));
    }
}

File: src/main/java/com/verlumen/tradestream/backtesting/BacktestRunnerImpl.kt
package com.verlumen.tradestream.backtesting

import com.google.inject.Inject
import com.google.protobuf.InvalidProtocolBufferException
import com.verlumen.tradestream.strategies.StrategyManager
import com.verlumen.tradestream.ta4j.BarSeriesFactory
import org.ta4j.core.AnalysisCriterion
import org.ta4j.core.BarSeries
import org.ta4j.core.BaseTradingRecord
import org.ta4j.core.Strategy
import org.ta4j.core.TradingRecord
import org.ta4j.core.criteria.pnl.ProfitLossCriterion
import org.ta4j.core.criteria.pnl.ProfitLossRatioCriterion
import org.ta4j.core.criteria.pnl.ReturnCriterion
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * Implementation of BacktestRunner that evaluates trading strategies using Ta4j.
 */
internal class BacktestRunnerImpl
    @Inject
    constructor(
        private val barSeriesFactory: BarSeriesFactory,
        private val strategyManager: StrategyManager,
    ) : BacktestRunner {
        @Throws(InvalidProtocolBufferException::class)
        override fun runBacktest(request: BacktestRequest): BacktestResult {
            require(request.candlesList.isNotEmpty()) { "Bar series cannot be empty" }

            val series = barSeriesFactory.createBarSeries(request.candlesList)
            val strategy =
                strategyManager.createStrategy(
                    series,
                    request.strategy.type,
                    request.strategy.parameters,
                )

            // Run the strategy
            val tradingRecord = runStrategy(series, strategy)

            // Calculate basic metrics
            val cumulativeReturn = calculateMetric(series, tradingRecord, ProfitLossCriterion())
            val profitFactor = calculateMetric(series, tradingRecord, ProfitLossRatioCriterion())
            val annualizedReturn = calculateAnnualizedReturn(series, tradingRecord)

            // Calculate risk metrics
            val volatility = calculateVolatility(series)
            val maxDrawdown = calculateMaxDrawdown(series)

            // Trade statistics
            val numberOfTrades = tradingRecord.positions.size
            val winRate = calculateWinRate(tradingRecord)
            val averageTradeDuration = calculateAverageTradeDuration(tradingRecord)

            // Risk-adjusted returns
            val sharpeRatio = calculateSharpeRatio(cumulativeReturn, volatility)
            val sortinoRatio = calculateSortinoRatio(series, tradingRecord)

            // Alpha/Beta (simplified calculation)
            val alpha = 0.0 // TODO: Implement when benchmark data is available
            val beta = 1.0 // TODO: Implement when benchmark data is available

            val score = calculateScore(sharpeRatio, maxDrawdown, winRate, annualizedReturn, profitFactor)

            return BacktestResult.newBuilder()
                .setCumulativeReturn(cumulativeReturn)
                .setAnnualizedReturn(annualizedReturn)
                .setSharpeRatio(sharpeRatio)
                .setSortinoRatio(sortinoRatio)
                .setMaxDrawdown(maxDrawdown)
                .setVolatility(volatility)
                .setWinRate(winRate)
                .setProfitFactor(profitFactor)
                .setNumberOfTrades(numberOfTrades)
                .setAverageTradeDuration(averageTradeDuration)
                .setAlpha(alpha)
                .setBeta(beta)
                .setStrategyScore(score)
                .build()
        }

        private fun runStrategy(
            series: BarSeries,
            strategy: Strategy,
        ): TradingRecord {
            val tradingRecord = BaseTradingRecord()

            // Skip unstable period at the start
            for (i in strategy.unstableBars until series.barCount) {
                // Check if we should enter long position
                if (strategy.shouldEnter(i)) {
                    // Enter with a position size of 1 unit
                    tradingRecord.enter(i, series.getBar(i).closePrice, series.numOf(1))
                }
                // Check if we should exit an open position
                else if (strategy.shouldExit(i) && tradingRecord.currentPosition.isOpened) {
                    tradingRecord.exit(i, series.getBar(i).closePrice, series.numOf(1))
                }
            }

            return tradingRecord
        }

        private fun calculateMetric(
            series: BarSeries,
            record: TradingRecord,
            criterion: AnalysisCriterion,
        ): Double = criterion.calculate(series, record).doubleValue()

        private fun calculateVolatility(series: BarSeries): Double {
            if (series.barCount < 2) {
                return 0.0
            }

            val returns =
                (1 until series.barCount).map { i ->
                    val previousClose = series.getBar(i - 1).closePrice
                    val currentClose = series.getBar(i).closePrice
                    currentClose.minus(previousClose)
                        .dividedBy(previousClose)
                        .doubleValue()
                }

            // Calculate standard deviation
            val mean = returns.average()
            val variance = returns.map { (it - mean).pow(2) }.average()

            return sqrt(variance)
        }

        private fun calculateMaxDrawdown(series: BarSeries): Double {
            if (series.barCount == 0) {
                return 0.0
            }

            var maxDrawdown = 0.0
            var peak = series.getBar(0).closePrice.doubleValue()

            (1 until series.barCount).forEach { i ->
                val price = series.getBar(i).closePrice.doubleValue()

                if (price > peak) {
                    peak = price
                }

                val drawdown = (peak - price) / peak
                maxDrawdown = max(maxDrawdown, drawdown)
            }

            return maxDrawdown
        }

        private fun calculateWinRate(record: TradingRecord): Double {
            if (record.positions.isEmpty()) {
                return 0.0
            }

            val winningTrades =
                record.positions
                    .count { it.isClosed && it.profit.isPositive }

            return winningTrades.toDouble() / record.positions.size
        }

        private fun calculateAverageTradeDuration(record: TradingRecord): Double {
            val closedPositions = record.positions.filter { it.isClosed }
            if (closedPositions.isEmpty()) {
                return 0.0
            }

            val totalDuration =
                closedPositions.sumOf {
                    it.exit.index - it.entry.index
                }

            return totalDuration.toDouble() / closedPositions.size
        }

        private fun calculateAnnualizedReturn(
            series: BarSeries,
            record: TradingRecord,
        ): Double {
            val totalReturn = calculateMetric(series, record, ReturnCriterion())
            val barsPerYear = 252 * 1440 // Assuming 1-minute bars and 252 trading days
            val years = series.barCount.toDouble() / barsPerYear

            // Use compound annual growth rate formula
            return (1 + totalReturn).pow(1 / years) - 1
        }

        private fun calculateSharpeRatio(
            returns: Double,
            volatility: Double,
        ): Double {
            val riskFreeRate = 0.02 // Assume 2% risk-free rate
            return if (volatility == 0.0) 0.0 else (returns - riskFreeRate) / volatility
        }

        private fun calculateSortinoRatio(
            series: BarSeries,
            record: TradingRecord,
        ): Double {
            // Simplified Sortino calculation using only negative returns
            val negativeReturns =
                (1 until series.barCount)
                    .map { i ->
                        val previousPrice = series.getBar(i - 1).closePrice.doubleValue()
                        val currentPrice = series.getBar(i).closePrice.doubleValue()
                        (currentPrice - previousPrice) / previousPrice
                    }
                    .filter { it < 0 }

            if (negativeReturns.isEmpty()) {
                return 0.0
            }

            // Calculate downside deviation
            val meanNegativeReturn = negativeReturns.average()
            val downsideDeviation =
                sqrt(
                    negativeReturns.map { (it - meanNegativeReturn).pow(2) }.average(),
                )

            // Calculate Sortino ratio using total return
            val totalReturn = calculateMetric(series, record, ReturnCriterion())
            val riskFreeRate = 0.02 // Assume 2% risk-free rate

            return if (downsideDeviation == 0.0) 0.0 else (totalReturn - riskFreeRate) / downsideDeviation
        }

        private fun calculateScore(
            sharpeRatio: Double,
            maxDrawdown: Double,
            winRate: Double,
            annualizedReturn: Double,
            profitFactor: Double,
        ): Double =
            0.25 * normalize(sharpeRatio) + // Risk-adjusted returns
                0.20 * (1 - maxDrawdown) + // Capital preservation
                0.20 * winRate + // Consistency
                0.20 * normalize(annualizedReturn) + // Absolute returns
                0.15 * normalize(profitFactor) // Profit efficiency

        private fun normalize(value: Double): Double {
            // Simple min-max normalization with reasonable bounds
            val min = -1.0
            val max = 2.0
            return ((value - min) / (max - min)).coerceIn(0.0, 1.0)
        }
    }
